### PR TITLE
feat(compiler): validate directive locations + undefined directives

### PR DESCRIPTION
**Pull request by lrlna**
partial of #246

This looks over each directive usage and checks whether it's in a correct location as specified by its original DirectiveDefinition. In case of incorrect usage `UnsupportedLocation` error is returned. Since we are already looking for the original DirectiveDefinition, this PR also adds a `UndefinedDefinition` error for undefined directives in case that DirectiveDefinition is not found.

While working on this validation rule, I realised we need to tidy up our ValidationDatabase before we implement any follow up rules. Primarily we are using two different paths for validating: `check_*` and `validate_*`, which should be unified in some way. 
We are also creating a lot of `let diagnostics = Vec::new()` for pretty much each validation step, and it would be good to not keep allocating new vecs for each node as we add diagnostics.

### CODE REVIEW

**Comment by lrlna** [comment]
sorry for 5000+ line additions! i swear most of these are tests 😅

File: crates/apollo-compiler/src/validation/validation_db.rs [hunk 1]
```diff
@@ -135,50 +156,82 @@ pub fn check_union_type_definition(
...
+    input_obj: Arc<hir::InputObjectTypeDefinition>,
 ) -> Vec<ApolloDiagnostic> {
     let mut diagnostics = Vec::new();
 
-    // TODO: validate extensions
-    for directive in schema_def.directives() {
-        diagnostics.extend(db.check_directive(directive.clone()));
-    }
+    diagnostics.extend(db.check_input_values(
+        (*input_obj).clone().input_fields_definition,
```
Comment (goto-bus-stop) [inline crates/apollo-compiler/src/validation/validation_db.rs:188]: only need to clone the inner thing
```suggestion
        input_obj.input_fields_definition.clone(),
```

**Review by goto-bus-stop — APPROVED** [review]
> We are also creating a lot of `let diagnostics = Vec::new()` for pretty much each validation step, and it would be good to not keep allocating new vecs for each node as we add diagnostics.

I think in this tree-walk design, it's almost certainly going to be faster with standalone functions that can borrow hir nodes and push into one shared diagnostics vec, even if it has to rerun the entire validation any time something changes. (compared to having the many cacheable queries with owned values everywhere). maaaybe it wouldn't make as much of a difference if we wrap everything in Arcs but i don't really have a good idea of how expensive atomic operations are.

this looks good tho 😄

**Comment by lrlna** [comment]
> > We are also creating a lot of `let diagnostics = Vec::new()` for pretty much each validation step, and it would be good to not keep allocating new vecs for each node as we add diagnostics.
> 
> I think in this tree-walk design, it's almost certainly going to be faster with standalone functions that can borrow hir nodes and push into one shared diagnostics vec, even if it has to rerun the entire validation any time something changes. (compared to having the many cacheable queries with owned values everywhere). maaaybe it wouldn't make as much of a difference if we wrap everything in Arcs but i don't really have a good idea of how expensive atomic operations are.
> 
> this looks good tho 😄

yea definitely something for us to think about once we simplify this validation a little bit!

**Pull request merged**